File: cpp/15.3Sum.cpp
/*
2016-8-20 17:10:32

https://leetcode.com/problems/3sum/

Given an array S of n integers, are there elements a, b, c in S such that a + b + c = 0? Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets.

For example, given array S = [-1, 0, 1, 2, -1, -4],

A solution set is:
[
  [-1, 0, 1],
  [-1, -1, 2]
]

*/

#include <iostream>
#include <algrithm>
using namespace std;

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
		
		if (nums.size() < 3)
		{
			return res;
		}
		
		sort(nums.begin(), nums.end());
		
		for (auto iter = nums.begin(); iter != nums.end() - 2 && *iter <= 0; ++ iter)
		{
			if (iter != nums.begin() && *iter == *(iter - 1))//防止出现相同的tuple；
				continue;
			
			auto left = iter + 1;
			auto right = nums.end() -1;
			
			while (right > left)
			{
				
				if (*iter + *left + *right == 0)
				{
					res.push_back({*iter, *left, *right});
					++ left;
					-- right;
					while (*left == *(left -1) && left < right)//防止出现相同的tuple；
						++ left;
					while (*right == *(right + 1) && left < right)
						-- right;
				}
				else if (*iter + *left + *right < 0)
				{
					++ left;
				}
				else
				{
					-- right;
				}
			}
			
		}
		
		return res;
    }
};

File: java/src/main/java/club/cser/leetcode/15. 3Sum.java
package club.cser.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.IntStream;

class ThreeSum {
    public List<List<Integer>> threeSum(int[] nums) {

        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; ++ i) {
            int left = i + 1;
            int right = nums.length - 1;

            // 重复数字 跳过
            if (i > 0 && nums[i] == nums[i - 1])
                continue;

            // 为正数的话 后面2个更大的数再加的话 肯定比0大
            if (nums[i] > 0)
                break;

            while (left < right) {
                if (nums[i] + nums[left] + nums[right] == 0) {
                    res.add(Arrays.asList(nums[i], nums[left], nums[right]));

                    ++ left;
                    -- right;

                    while (left < right && nums[left - 1] == nums[left]) {
                        ++ left;
                    }

                    while (left < right && nums[right + 1] == nums[right]) {
                        -- right;
                    }
                } else if (nums[i] + nums[left] + nums[right] > 0) {
                    -- right;
                } else {
                    ++ left;
                }
            }
        }

        return res;
    }
}